src/cliques.c

```diff
@@ -0,0 +1,628 @@
+//
+// Copyright (C) 2020 - Julius Jonusas 
+//
+// This file is free software, see the digraphs/LICENSE.
+
+
+// C headers
+#include <setjmp.h>   // for longjmp, setjmp, jmp_buf
+#include <stdbool.h>  // for true, false, bool
+#include <stddef.h>   // for NULL
+#include <stdint.h>   // for uint16_t, uint64_t
+#include <stdlib.h>   // for malloc, NULL
+
+// GAP headers
+#include "src/compiled.h"
+
+// Digraphs package headers
+#include "bitarray.h"         // for BitArray
+#include "conditions.h"       // for Conditions
+#include "digraphs-debug.h"   // for DIGRAPHS_ASSERT
+#include "homos-graphs.h"     // for Digraph, Graph, . . .
+#include "perms.h"            // for MAXVERTS, UNDEFINED, PermColl, Perm
+
+////////////////////////////////////////////////////////////////////////////////
+// Macros
+////////////////////////////////////////////////////////////////////////////////
+
+#define MIN(a, b) (a < b ? a : b)
+
+////////////////////////////////////////////////////////////////////////////////
+// Forward declarations
+////////////////////////////////////////////////////////////////////////////////
+
+// Defined in digraphs.h
+Int DigraphNrVertices(Obj);
+Obj FuncOutNeighbours(Obj, Obj);
+Obj FuncADJACENCY_MATRIX(Obj, Obj);
+
+// GAP level things, imported in digraphs.c
+extern Obj IsDigraph;
+extern Obj Infinity;
+extern Obj IsSymmetricDigraph;
+extern Obj GeneratorsOfGroup;
+extern Obj AutomorphismGroup;
+extern Obj IsPermGroup;
+extern Obj IsDigraphAutomorphism;
+extern Obj LargestMovedPointPerms;
+extern Obj SmallestMovedPointPerm;
+extern Obj IsClique;
+extern Obj IsTrivial;
+extern Obj Orbit;
+extern Obj Stabilizer;
+extern Obj IsSubset;
+extern Obj OnTuples;
+extern Obj Group;
+extern Obj ClosureGroup;
+
+////////////////////////////////////////////////////////////////////////////////
+// Global variables
+////////////////////////////////////////////////////////////////////////////////
+
+static Obj GAP_FUNC;       // Variable to hold a GAP level hook function
+
+static Obj (*HOOK)(void*,  // HOOK function applied to every homo found
+                   const BitArray*,
+                   const uint16_t);
+static void* USER_PARAM;   // A USER_PARAM for the hook
+
+static jmp_buf OUTOFHERE;  // So we can jump out of the deepest
+
+static Graph* GRAPH;       // Graphs to hold incoming GAP symmetric digraphs
+
+static BitArray*  CLIQUE;
+static Conditions*  TRY;
+static Conditions*  BAN;
+
+static BitArray* ORB_LOOKUP;          // points in orbit
+static Obj ORBIT;
+
+////////////////////////////////////////////////////////////////////////////////
+// Hook functions
+////////////////////////////////////////////////////////////////////////////////
+
+static Obj clique_hook_collect(void* user_param,
+                               const BitArray* clique,
+                               const uint16_t nr) {
+  UInt   i;
+  Obj    c;
+
+  if (TNUM_OBJ((Obj) user_param) == T_PLIST_EMPTY) {
+    RetypeBag(user_param, T_PLIST);
+  }
```

Comment:
This shouldn't be necessary (and I would recommend avoiding it!) as you use `ASS_LIST` below (good!)